text/0000-std-rpc-api.md

```diff
@@ -0,0 +1,393 @@
+- Issue: [Unity-Technologies/com.unity.multiplayer#0000](https://github.com/Unity-Technologies/com.unity.multiplayer/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Existing MLAPI [**Convenience** and **Performance** RPCs](https://mlapi.network/wiki/messaging-system/) are sub-optimal in terms of required boilerplate code, API design and runtime performance.
+
+This RFC proposes a new **Standard** RPC API that excels at minimal boilerplate, API design/clarity, extensibility (future-proofing) and runtime performance.
+
+The initial implementation of this standard would be kept minimal and further-extended with additional features down the line as it will also support future extensions and enhancements without changing the standard API necessarily.
+
+# Motivation
+[motivation]: #motivation
+
+Achieve better performance, unify & standardize RPC API, make RPC API extensible & future-proof and reduce boilerplate code.
+
+Also a simple internal performance benchmark came with these results:
+
+    Local RPC Send Performance Test (524280 calls)
+    Convenience -> 4361ms
+    Performance -> 2607ms
+    Standard -> 2205ms
+    Local RPC Recv Performance Test (524280 calls)
+    Convenience -> 9626ms
+    Performance -> 5923ms
+    Standard -> 1935ms
+
+We're looking at **up to 5.24x better performance** (compared to convenience API vs standard API receive performance).
+
+Beyond that, on an internal project, we saw roughly **%80 less boilerplate code** compared to performance RPCs and standard RPCs which is also a huge win.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The multiplayer framework provides 2 main network constructs ([ServerRpc](#serverrpc) and [ClientRpc](#clientrpc)) to execute logic on either server-side or client-side. This concept is often called [Remote Procedure Call (RPC)](https://en.wikipedia.org/wiki/Remote_procedure_call) and has wide adoption across the industry.
+
+## RPC Methods
+
+A framework user (Unity developer) can declare multiple RPCs under a `NetworkBehaviour` and inbound/outbound RPC calls will be replicated as a part of its replication in a network frame.
+
+A method turned into an RPC is no longer a regular method, it will have its  own implications on direct calls and in the network pipeline (see [Execution Table](#execution-table)).
+
+### ServerRpc
+
+A `ServerRpc` can be invoked by a client to be executed on the server.
+
+Developer can declare a `ServerRpc` by marking a method with `[ServerRpc]` attribute and making sure to have `ServerRpc` suffix in the method name.
+
+```cs
+[ServerRpc]
+void PingServerRpc(int framekey) { /* ... */ }
+```
+
+Developer can invoke a `ServerRpc` by making a direct function call with parameters:
+
+```cs
+void Update()
+{
+    if (Input.GetKeyDown(KeyCode.P))
+    {
+        PingServerRpc(Time.frameCount); // Client -> Server
```

Comment:
What happens when i connect to a server which has a different network signature? E.g. my client uses an older version where this was an `int`, the server is up to date and this uses a `string`. 

How does the server handle this case? Is the entire RPC dropped?

Reply:
I'd consider RPC-stuff a "distributed API", kind of like REST. For games, **usually** you'll have enough control to make sure clients and servers are on the same version. Ever had a game (or app) tell you "you need to update to play" or "you need to update to play _online_"? That's why.

If you really need to support different versions talking to each other, I'd first check if you **really** need to support different versions. By supporting this, you're making an already extremely complex thing even more complex (with lots of really nasty bugs you might run into that will be a nightmare to isolate, reproduce and fix).

So most likely, if you've done it right, the server will kick the client if the versions don't match, problem solved. This, btw, also helps with cracked versions (that often are outdated). It can also help, to a certain degree, with clients that would cheat.

If you still really need to support different versions working together: Include version numbers in your method names. In all of them. And tag old method names with Obsolete to make builds fail when still accessing them. In that case, it's kind of like versioning for REST-APIs, basically "URI versioning", like here: https://restfulapi.net/versioning/

But most likely, you will regret having made that choice that after about a week ;-)

Reply:
cross-build and cross-version compatibility is something we need to discuss separately. having said that, MLAPI doesn't support cross-build and/or cross-version compatibility today.

Reply:
Cross-build compatibility is not supported? To me, that would mean that if you do a server-build, a build for Windows and a build for Android, those could not talk to each other. That doesn't sound plausible, so I assume you mean something else with "cross-build"? But then, MLAPI would also support "cross-version" as long as people don't mess with the interfaces / method signatures.

Ah ... maybe that's it: So MLAPI does not support two builds talking to each other, if there was any change in any of the methods that belong to the RPC-contract? That would make total sense to me.

Is that documented somewhere? It's a limitation that makes total sense - but it's also a limitation that people really need to be aware of. My understanding would be that any change to any RPC would break compatibility but Custom Messages should be fine.

Reply:
Yep, this:

> Ah ... maybe that's it: So MLAPI does not support two builds talking to each other, if there was any change in any of the methods that belong to the RPC-contract? That would make total sense to me.

I will try to inline this conversation too.

Reply:
The MLAPI does support different builds talking to each other. They just need to have the same RPC name.

Reply:
> This is all because of the deterministic hashing of RPC names.

I think this is a very important thing to keep. @MFatihMAR The mapping of RPCs being non-deterministic (i.e. one change could make the whole thing break) seems like an implementation choice that will be easy to regret. It does seem to me that this would be an implementation-level issue (not a design issue) but I don't know enough about the "magic" that would wire up RPCs with IDs.

Reply:
@JashanChittesh

Deterministic hashing of RPC names means just hash the method name, that's not a great idea, your method signature could change therefore you actually need to change your serialization code and not resolve into same RPC serialization code. Having said that, we will be backward-compatible as much as possible, [I inlined some more details here](https://github.com/Unity-Technologies/com.unity.multiplayer.rfcs/blob/rfc/std-rpc-api/text/0000-std-rpc-api.md#backward-compatibility).

It is fundamentally:

> Framework registers RPC methods statically by their deterministic [method signature](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/methods#method-signatures) hash which means as long as method signature stays the same, backward-compatibility will be possible but if method signature changes, same method with the same name under same namespace and type will not be compatible with its predecessor.

I will also inline other cross-compatibility sections very soon, thanks for raising this up!